### PR TITLE
Introduce string map class and switch to it

### LINKED ISSUE #4763: ordered_map<cstring, ...> fiasko
These maps are used at least in two important places: 
  - To jold declarations incide `IR::IndexedVector` (so we do have `ordered_map<cstring, IDeclaration*>` there)
  - To store guts of Util::JsonObject (we are using `ordered_map<cstring, IJson*>`)

And it turns out `ordered_map<cstring, something>` is a very bad data structure essentially defeating both the purpose of map and purpose of `cstring`.

Under the hood `ordered_map<cstring, V>` is a pair of of `std::list<std::pair<cstring, V>>` and `std::map<cstring*, list::iterator, Comp>`. So every time we insert something into an `ordered_map`, the value is appended to list, and corresponding map entry is pointed to corresponding iterator. So, in the end we're having map-like interface for lookup, but iteration order that corresponds to the insertion order. The devil is in the details:

1. It does not make any sense to store `cstring*` as a key. cstrings are cheap to copy, unique, so additional level of indirection creates lots of malloc traffic for no particular reason.
2. It does not make any sense to use `std::map` actually. Yes, the intention is to allow `lower_bound` / `upper_bound`, etc. interface. But it seems that in the whole codebase noone uses it. `std::map` itself is quite expensive in terms of overhead (see https://github.com/p4lang/p4c/pull/4150 for a similar story), plus each entry it separately allocated, again creating huge malloc traffic.
3. Last, but not the least. The comparator `Comp` mentioned above essentially dereferences keys and calls `std::less<cstring>`. However, the latter calls `strcmp` under the hood. So, inserting something into `ordered_map<cstring, V>` causes *lexicographical string comparison* for the string values it is holding defeating all `cstring` advantages...

Certainly the solution is to replace `std::map` into much more reliable variant. We really just use unordered map there and this'd solve the majority of issues.

### PR DESCRIPTION
**Pull request by asl**
The main change of this PR is a new `string_map<V>` class that does `ordered_map<cstring, V>` but in a proper way:
 - Key are stored as cstrings in the underlying abseil hash map.
 - Heterogenous lookup (with `std::string_view` keys) is supported. Special care is done not to create cstrings in case if they are not in the map already
 - For now map is ordered, values are stored in `std::list`, similar to `ordered_map`. Likely we'd need to allow `std::vector` / `absl::inlined_vector` as storage as well via dedicated policy, but I need to collect more statistics on this
 - Several ordered_map construction bugs were fixed as well

Give this class, I moved two main users of `ordered_map<cstring, V>` to it:
 - IDeclaration
 - JSonObject
trying to refactor & modernize the API to allow & use heterogenous lookup as well.

While there I checked all instances of `std::map<cstring, something` in frontend / midend and replaced them with unordered maps where appropriate.

While it might look like a breaking change, it is not. Even more, the code eases adopting of downstream code to explicit `const char* => cstring` conversion, as heterogenous lookup makes these conversions unnecessary. I moved code in backends to this where appropriate.

Fixes #4763

### CODE REVIEW

**Comment by asl** [comment]
For the downstream testcase the compile time changes from:
```
Executed in   70.79 secs    fish           external
   usr time  107.65 secs    0.05 millis  107.65 secs
   sys time    3.89 secs    1.44 millis    3.89 secs
```
down to:
```
Executed in   63.48 secs    fish           external
   usr time   81.99 secs    0.09 millis   81.99 secs
   sys time    3.71 secs    5.54 millis    3.70 secs
```

So we are having nice 10% speedup. Two things contribute to this change:
 - Removal of strcmp in IDeclaration's `ordered_map`
 - Much less malloc traffic there.

File: lib/string_map.h [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+#include <initializer_list>
+#include <list>
+#include <stdexcept>
+#include <utility>
+
+#include "absl/container/flat_hash_map.h"
+#include "cstring.h"
+
+// Map is ordered by order of element insertion.
```
Comment (fruffy) [inline lib/string_map.h:26]: Can you add a more detail comment on when to use this map or what is purpose is? Similar to the description in the PR.

**Comment by vlstill** [comment]
> For now map is ordered, values are stored in std::list, similar to ordered_map. Likely we'd need to allow std::vector / absl::inlined_vector as storage as well via dedicated policy, but I need to collect more statistics on this

Or we could just always use `std::vector`. I doubt `std::list` will ever be better. Alternatively, if the iteration is rare, some sort of intrusive (singly-) linked list might be usable (I hope noone is reverse iterating over these).

File: lib/string_map.h [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    const_reverse_iterator rend() const noexcept { return data.rend(); }
+    const_iterator cbegin() const noexcept { return data.cbegin(); }
+    const_iterator cend() const noexcept { return data.cend(); }
+    const_reverse_iterator crbegin() const noexcept { return data.crbegin(); }
+    const_reverse_iterator crend() const noexcept { return data.crend(); }
+
+    bool empty() const noexcept { return data.empty(); }
+    size_type size() const noexcept { return data_map.size(); }
+    size_type max_size() const noexcept { return data_map.max_size(); }
+    bool operator==(const string_map &a) const { return data == a.data; }
```
Comment (vlstill) [inline lib/string_map.h:102]: There should probably be a comment that two `string_map`s are equal even if their iteration order differs. This is probably OK to have it implemented that way, but one should not need to go into source to learn that.

File: lib/string_map.h [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    bool operator!=(const string_map &a) const { return data != a.data; }
+    void clear() {
+        data.clear();
+        data_map.clear();
+    }
+
+    // Functions above do have `std::string_view` versions. Here we are having
```
Comment (vlstill) [inline lib/string_map.h:109]: Probably "below" instead of "above"

File: lib/string_map.h [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+    // important special case: if `a` is not something that was interned, we do
+    // not copy / intern it, we know for sure that `a` is not in the map and we
+    // do not need to perform a lookup
```
Comment (vlstill) [inline lib/string_map.h:112]: The comment should probably be on the `string_view` version and it should be Doxygen.

File: lib/string_map.h [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        if (key.isNull()) return false;
+
+        return data_map.contains(key);
+    }
+
+    template <typename Key>
+    V &operator[](Key &&k) {
+        auto it = find(key_type(std::forward<Key>(k)));
+        if (it == data.end()) {
+            it = data.emplace(data.end(), k, V());
```
Comment (vlstill) [inline lib/string_map.h:148]:
```suggestion
            it = data.emplace(data.end(), std::piecewise_construct, std::forward_as_tuple(k), std::forward_as_tuple());
```

Yes, this is ugly, but it prevents creating a temporary that is then copied to the constructor.

File: lib/string_map.h [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+        auto it = find(std::forward<Key>(k));
+        if (it == data.end()) throw std::out_of_range("string_map::at");
+        return it->second;
+    }
+
+    template <typename Key, typename... VV>
+    std::pair<iterator, bool> emplace(Key &&k, VV &&...v) {
+        auto it = find(key_type(std::forward<Key>(k)));
+        if (it == data.end()) {
+            it = data.emplace(data.end(), std::piecewise_construct_t(), std::forward_as_tuple(k),
```
Comment (vlstill) [inline lib/string_map.h:171]:
```suggestion
            it = data.emplace(data.end(), std::piecewise_construct, std::forward_as_tuple(k),

File: lib/string_map.h [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+                              std::forward_as_tuple(std::forward<VV>(v)...));
+            data_map.emplace(it->first, it);
+            return std::make_pair(it, true);
```
Comment (vlstill) [inline lib/string_map.h:174]: just `std::pair` (or maybe even `{it, true}`) should be sufficent.

File: lib/string_map.h [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+        }
+        return std::make_pair(it, false);
```
Comment (vlstill) [inline lib/string_map.h:176]: And the same here.

**Review by vlstill — COMMENTED** [review]
For now, I went through the data structure implementation. Thank you for your contribution!

**Comment by asl** [comment]
> Or we could just always use `std::vector`. I doubt `std::list` will ever be better. Alternatively, if the iteration is rare, some sort of intrusive (singly-) linked list might be usable (I hope noone is reverse iterating over these).

Yes. `std::list` definitely has much more malloc traffic overhead. And this case is certainly optimized for often erasures. For `IDeclaration` looks like an intrusive vector would be preferable. For JSonObject – likely the usual std::vector. Anyway, this is something I am planning to measure and play with :)

File: lib/string_map.h (hunk 5 above)
Comment (asl) [reply lib/string_map.h]: Yeah, nice catch. Completely forgot this here.

File: lib/string_map.h (hunk 7 above)
Comment (asl) [reply lib/string_map.h]: Yeah. It's just muscle memory :)

File: lib/string_map.h (hunk 2 above)
Comment (asl) [reply lib/string_map.h:116]: This is same as with `ordered_map`. So, the insertion order does matter. I thought it was ok assumption provided that there is already an `ordered_map` that is (over-)widely used across the codebase. Comment added.

File: lib/string_map.h (hunk 1 above)
Comment (asl) [reply lib/string_map.h]: Done!

File: lib/string_map.h [hunk 9]
```diff
@@ -23,7 +23,20 @@ limitations under the License.
 #include "absl/container/flat_hash_map.h"
 #include "cstring.h"
 
-// Map is ordered by order of element insertion.
+/// @file
+/// Map with string keys that is ordered by order of element insertion.
+/// Use this map only when stable teration order is significant. For everything
```
Comment (vlstill) [inline lib/string_map.h:28]:
```suggestion
/// Use this map only when stable iteration order is significant. For everything

File: lib/string_map.h [hunk 10]
```diff
@@ -23,7 +23,20 @@ limitations under the License.
...
+/// else consider using some other unordered containers with cstring keys. Never
+/// use std::map<cstring, ...> or ordered_map<cstring, ...> as these maps will
+/// have significant overhead due to lexicographical ordering of the keys.
+///
+/// For this particular implementation:
+///  * Key are stored as cstrings in the underlying abseil hash map.
+///  * Heterogenous lookup (with std::string_view keys) is supported. Special
+///    care is done not to create cstrings in case if they are not in the map
+///    already
+//   * Values are stored in std::list, similar to ordered_map.
```
Comment (vlstill) [inline lib/string_map.h:38]:
```suggestion
///  * Values are stored in std::list, similar to ordered_map.

File: lib/string_map.h [hunk 11]
```diff
@@ -23,7 +23,20 @@ limitations under the License.
 #include "absl/container/flat_hash_map.h"
 #include "cstring.h"
 
-// Map is ordered by order of element insertion.
+/// @file
```
Comment (vlstill) [inline lib/string_map.h:26]: Maybe associate the comment with the class instead? I'm not sure what best practices for doxygen we have, but it seems to me it would make the comment more visible.

File: lib/string_map.h [hunk 12]
```diff
@@ -99,18 +111,20 @@ class string_map {
     bool empty() const noexcept { return data.empty(); }
     size_type size() const noexcept { return data_map.size(); }
     size_type max_size() const noexcept { return data_map.max_size(); }
+    /// As the map is ordered, two maps are only considered equal if the elements
+    /// are inserted in the same order. Same as with ordered_map though
```
Comment (vlstill) [inline lib/string_map.h:115]:
```suggestion
    /// are inserted in the same order. Same as with ordered_map though.

File: lib/string_map.h (hunk 2 above)
Comment (vlstill) [reply lib/string_map.h:116]: Yeah, I got `data` and `data_map` confused.

**Review by vlstill — COMMENTED** [review]
Few nitpicks in the map implementation.

File: frontends/p4/checkNamedArgs.cpp [hunk 13]
```diff
@@ -21,7 +21,7 @@ namespace P4 {
 bool CheckNamedArgs::checkArguments(const IR::Vector<IR::Argument> *arguments) {
     bool first = true;
     bool hasName = false;
-    std::map<cstring, const IR::Argument *> found;
+    absl::flat_hash_map<cstring, const IR::Argument *> found;
```
Comment (vlstill) [inline frontends/p4/checkNamedArgs.cpp:24]: Should we use the hasher from `Utils`?

File: ir/v1.def [hunk 14]
```diff
@@ -435,6 +435,7 @@ class AttribLocals : ISimpleNamespace {
     Util::Enumerator<IDeclaration> *getDeclarations() const override {
         return locals.valueEnumerator()->as<const IDeclaration *>(); }
     IDeclaration getDeclByName(cstring name) const override { return locals[name]; }
+    IDeclaration getDeclByName(std::string_view name) const override { return locals[cstring(name)]; }
```
Comment (vlstill) [inline ir/v1.def:438]: Do we want this overload if it just casts to `cstring` anyway? It is probably good for consistency, but could lead to an unexpected cast.

File: lib/cstring.cpp
Comment (vlstill) [inline lib/cstring.cpp:1]: Nice! Might be nicer to split this change to a separate PR as it does not seem to be connected to the rest. Or at least mention it in the PR description that cstring's cache representation was changed.

File: lib/json.h [hunk 15]
```diff
@@ -55,32 +55,36 @@ class JsonValue final : public IJson {
...
+    JsonValue(int v) : tag(Kind::Number), value(v) {}            // NOLINT
+    JsonValue(long v) : tag(Kind::Number), value(v) {}           // NOLINT
+    JsonValue(long long v);                                      // NOLINT
+    JsonValue(unsigned v) : tag(Kind::Number), value(v) {}       // NOLINT
+    JsonValue(unsigned long v) : tag(Kind::Number), value(v) {}  // NOLINT
+    JsonValue(unsigned long long v);                             // NOLINT
+    JsonValue(double v) : tag(Kind::Number), value(v) {}         // NOLINT
+    JsonValue(float v) : tag(Kind::Number), value(v) {}          // NOLINT
+    JsonValue(cstring s) : tag(Kind::String), str(s) {}          // NOLINT
+    // FIXME: replace these two ctors with std::string view, cannot do now as
```
Comment (vlstill) [inline lib/json.h:69]:
```suggestion
    // FIXME: replace these two ctors with std::string_view, cannot do now as

File: test/gtest/cstring.cpp [hunk 16]
```diff
@@ -143,4 +143,10 @@ TEST(cstring, literalSuffix) {
     EXPECT_TRUE((std::is_same_v<cstring, decltype(""_cs)>));
 }
 
+TEST(cstring, is_cached) {
+    EXPECT_FALSE(
+        cstring::is_cached("we really do not expect that this string is already in cstring cache"));
+    EXPECT_TRUE(cstring::is_cached("test"));
```
Comment (vlstill) [inline test/gtest/cstring.cpp:149]: We should create the `cstring` in this test so that it can be run separately.

File: test/gtest/string_map.cpp [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+namespace Test {
+
+TEST(StringMap, RangeConstructor) {
+    string_map<unsigned>::value_type myMap[] = {{"One"_cs, 1},   {"One"_cs, 2},   {"One"_cs, 3},
+                                                {"Two"_cs, 1},   {"Two"_cs, 2},   {"Two"_cs, 3},
+                                                {"Three"_cs, 1}, {"Three"_cs, 2}, {"Three"_cs, 3}};
+
+    string_map<unsigned> first(std::begin(myMap), std::end(myMap));
+    EXPECT_THAT(first, ElementsAre(std::make_pair("One"_cs, 1), std::make_pair("Two"_cs, 1),
+                                   std::make_pair("Three"_cs, 1)));
```
Comment (vlstill) [inline test/gtest/string_map.cpp:38]:
```suggestion
    EXPECT_THAT(first, ElementsAre(std::pair("One"_cs, 1), std::pair("Two"_cs, 1),
                                   std::pair("Three"_cs, 1)));

**Review by vlstill — COMMENTED** [review]
A few comments for the rest of the code. Once we have C++20, we should try to revisit if we can replace the overloads that are the same for `cstring` and `std::string_view` with a single one taking a concept that is either. But I am not sure if GCC 9.2 has enough concepts support to allow that or if we would need to wait until we can drop it.

File: frontends/p4/checkNamedArgs.cpp (hunk 13 above)
Comment (asl) [reply frontends/p4/checkNamedArgs.cpp:24]: It will be used via our `std::hash` specialization. So, nothing special is needed. Though, abseil on top would salt the hash, so the iteration order will be randomized. I do not have preferences which approach is better.

File: ir/v1.def (hunk 14 above)
Comment (asl) [reply ir/v1.def:438]: Well, we really need this overload since it is an implementation of an interface, without it we will have pure virtual function. `locals` here is `NameMap` – another flavour of string map, which is used only in v1 IR. I decided not to touch it as I'm feeling we do not have enough test coverage for it.

File: frontends/p4/checkNamedArgs.cpp (hunk 13 above)
Comment (asl) [reply frontends/p4/checkNamedArgs.cpp:24]: After thinking about this a bit more, maybe we can adopt the following rule:
 - For local maps (within the method, etc) we can use `Util::Hash` directly. This way we won't have additional random salt on top of the existing hash, so the hashing latency will be reduced.
 - For members we'd stick with absel's randomized hashes. This way the iteration order will be different on each invocation. As a result, all possible accidentally added iterations could be noticed much earlier.

What do you think?

**Comment by asl** [comment]
@vlstill I factored out cstring bits into https://github.com/p4lang/p4c/pull/4780 to decompose the changes

File: lib/cstring.cpp
Comment (asl) [reply lib/cstring.cpp:1]: https://github.com/p4lang/p4c/pull/4780

**Comment by ChrisDodd** [comment]
> For the downstream testcase the compile time changes from:
> 
> ```
> Executed in   70.79 secs    fish           external
>    usr time  107.65 secs    0.05 millis  107.65 secs
>    sys time    3.89 secs    1.44 millis    3.89 secs
> ```
> 
> down to:
> 
> ```
> Executed in   63.48 secs    fish           external
>    usr time   81.99 secs    0.09 millis   81.99 secs
>    sys time    3.71 secs    5.54 millis    3.70 secs
> ```
> 
> So we are having nice 10% speedup. Two things contribute to this change:
> 
> * Removal of strcmp in IDeclaration's `ordered_map`
> * Much less malloc traffic there.

How does the performance compare to hvec_map?

**Comment by asl** [comment]
> How does the performance compare to hvec_map?

Its performance was lower in my initial tests. Checked once again to ensure we're comparing apples-to-apples replacing declaration map in `IndexedVector` with `hvec_map<cstring, IDeclaration*>`:

```
Executed in   65.31 secs    fish           external
   usr time  101.39 secs    0.10 millis  101.39 secs
   sys time    1.95 secs    2.28 millis    1.95 secs
```

**Comment by asl** [comment]
Rebased after https://github.com/p4lang/p4c/pull/4780

File: frontends/p4/checkNamedArgs.cpp (hunk 13 above)
Comment (asl) [reply frontends/p4/checkNamedArgs.cpp]: So, I implemented this approach for local maps. Also, found one "funny" case when we had map `cstring -> ID` which I replaces with set.

**Comment by asl** [comment]
@vlstill @fruffy Anything else left to address?

**Review by fruffy — APPROVED** [review]
LGTM
